### DISCUSSION TITLE
Converters with a custom ValueRenderer

### DISCUSSION THREAD

**Discussion by joshqou** [apple/pkl#322] [Q&A]
I want to abstract a configuration format that is not officially supported, however ValueRenderer doesn't seem to support the `converters` mapping.

```pkl
class Space {
    box: Box
}

class Box {
    id: String
    color: Color
    label: String?
}

space: Space
```

```ini
[Space]
BoxID=box_1
BoxColor=red
BoxLabel=A readable label
```

Is there a way of doing this using pkl instead of outsourcing the conversion and rendering to `jq`?

**Comment by bioball** [comment]
I'm curious what `jq` has to do with this; do you ultimately need JSON?

What format is this? Is this some flavor of INI?

Generally, you want to use a renderer to define how Pkl gets serialized to text. You can write an in-language renderer (see our [toml renderer](https://github.com/apple/pkl-pantry/blob/main/packages/pkl.toml/toml.pkl) for an example). @Madmegsox1 also submitted a PR to add an INI renderer to the stdlib in https://github.com/apple/pkl/pull/149, which might help you.

**Reply by joshqou** [reply]
> I'm curious what `jq` has to do with this; do you ultimately need JSON?

I've been using `jq` to do the actual rendering by having pkl render to JSON. I hadn't read the `output` section in the language ref initially so it made more sense to me

```sh
$ set -eu -o pipefail
$ pkl eval -f json file.pkl | jq -rf file.jq # > file.conf 
```

```jq
{
    Space: (if .space != null then .space | {
        BoxId: .box?.id,
        BoxColor: .box?.color,
        BoxLabel: .box?.label
    } else null end)
} | to_conf
```

> What format is this? Is this some flavor of INI?

It's a flavor of INI with multiple key assignment support, similar to `systemd.unit`.

> Generally, you want to use a renderer to define how Pkl gets serialized to text. You can write an in-language renderer (see our toml renderer for an example). @​Madmegsox1 also submitted a PR to add an INI renderer to the stdlib in #149, which might help you.

I did try writing the renderer in pkl however got caught out when I realised `converters` wasn't implemented by ValueRenderer since I wasn't sure how to implement `converters` myself within pkl. I don't think the PR renderer would work for the specific INI flavor I'm targetting as it's using `Map`

The OP may have been a bit vague so here's a more clear description of what I'm wanting to achieve within pkl:

- Target a custom INI flavor where a key may be assigned multiple times. I'm currently handling this inside my `to_conf` jq function
- Wanting to unflatten the schema. i.e: `box { id = "str" }` instead of `BoxId = "str"` for readability

**Reply by holzensp** [reply]
A general implementation of converters also needs the ability to "resume" conversion; at the moment, a _path_ converter (one where the key is `"foo.bar.baz"` instead of `Foo.getClass()`) can not be defined recursively, because the definition doesn't have access to the path itself.

That said, for this application, just defining your own output should be fine.
